src/main/java/co/cask/spanner/source/SpannerSource.java

```diff
@@ -0,0 +1,283 @@
+                                             for (Map.Entry<String, String> entry : configuration) {
+                                               config.put(entry.getKey(), entry.getValue());
+                                             }
+                                             return config;
+                                           }
+                                         }));
+
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    schema = config.getSchema();
+  }
+
+  @Override
+  public void transform(KeyValue<NullWritable, ResultSet> input, Emitter<StructuredRecord> emitter) throws Exception {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    List<Schema.Field> fields = schema.getFields();
+    ResultSet resultSet = input.getValue();
+    for (Schema.Field field : fields) {
+      String fieldName = field.getName();
+      Type columnType = resultSet.getColumnType(fieldName);
+      if (columnType == null || resultSet.isNull(fieldName)) {
+        continue;
+      }
+      switch (columnType.getCode()) {
+        case BOOL:
+          builder.set(fieldName, resultSet.getBoolean(fieldName));
+          break;
+        case INT64:
+          builder.set(fieldName, resultSet.getLong(fieldName));
+          break;
+        case FLOAT64:
+          builder.set(fieldName, resultSet.getDouble(fieldName));
+          break;
+        case STRING:
+          builder.set(fieldName, resultSet.getString(fieldName));
+          break;
+        case BYTES:
+          ByteArray byteArray = resultSet.getBytes(fieldName);
+          builder.set(fieldName, byteArray.toByteArray());
+          break;
+      }
+    }
+    emitter.emit(builder.build());
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    LOG.info("Run finished, closing spanner client");
+    // free up spanner resources
+    if (spanner != null) {
+      spanner.close();
+    }
+  }
+
+  private void initializeConfig(Configuration configuration) {
+    Map<String, String> configurationMap = new HashMap<>();
+    configurationMap.put(SpannerConstants.PROJECT_ID, config.project);
+    configurationMap.put(SpannerConstants.SERVICE_ACCOUNT_FILE_PATH, config.serviceAccountFilePath);
+    configurationMap.put(SpannerConstants.INSTANCE_ID, config.instance);
+    configurationMap.put(SpannerConstants.DATABASE, config.database);
+    configurationMap.put(SpannerConstants.QUERY, config.query);
+
+    configurationMap.entrySet().stream()
+      .filter(e -> e.getValue() != null)
+      .forEach(e -> configuration.set(e.getKey(), e.getValue()));
+  }
+
+  /**
+   * Serialize the object and store the serialized string in configuration identified by the propertyName
+   * @throws IOException
+   */
+  private void serializeObject(Object object, Configuration configuration, String propertyName) throws IOException {
```

Comment:
it's cleaner to make this method return a string and make the caller set the value in the Configuration object. Try to keep it so that methods do one thing.